firedrake/interpolation.py

```diff
@@ -1664,3 +1666,126 @@ def multHermitian(self, source_vec, target_vec):
+                         bcs=bcs, allow_missing_dofs=allow_missing_dofs)
+
+    @PETSc.Log.EventDecorator()
+    def interpolate(self, function, output=None, adjoint=False):
+        """
+        Compute the Clément interpolant applied to a source function.
+
+        Parameters
+        ----------
+        function: firedrake.function.Function or firedrake.cofunction.Cofunction
+                  Function to be interpolated.
+        output: firedrake.function.Function or firedrake.cofunction.Cofunction
+                A function to contain the output.
+        adjoint: bool
+                   Set to true to apply the adjoint of the interpolation
+                   operator.
+
+        Returns
+        -------
+        firedrake.function.Function or firedrake.cofunction.Cofunction
+            The resulting interpolated function.
+        """
+        from firedrake.assemble import assemble
+        if adjoint:
+            raise NotImplementedError("Adjoint of Clément interpolation not implemented.")
+        Vs = function.function_space()
+        element = Vs.ufl_element()
+        if not (element.family() == "Discontinuous Lagrange" and element.degree() == 0):
+            raise ValueError("Source function must live in P0 space.")
+        rank = len(Vs.value_shape)
+        if rank not in (0, 1, 2):
+            raise ValueError(f"Rank-{rank} tensors are not supported.")
+        if rank != len(self.V.value_shape):
+            raise ValueError(f"Rank-{rank} input inconsistent with target space.")
+        mesh = self.V.mesh()
+        dim = mesh.topological_dimension()
+        dX = ufl.dx(domain=mesh)
+        if output is None:
+            output = ffunc.Function(self.V)
+
+        # Compute volume of each element
+        volume = ffunc.Function(functionspace.FunctionSpace(mesh, "DG", 0))
+        volume.interpolate(ufl.CellVolume(mesh))
+
+        # Take weighted average
+        domain = {
+            0: "{[i]: 0 <= i < out.dofs}",
+            1: f"{{[i, j]: 0 <= i < out.dofs and 0 <= j < {dim}}}",
+            2: f"{{[i, j, k]: 0 <= i < out.dofs and 0 <= j < {dim} and 0 <= k < {dim}}}",
+        }[rank]
+        instructions = {
+            0: "out[i] = out[i] + vol[0] * f[0]",
+            1: "out[i, j] = out[i, j] + vol[0] * f[0, j]",
+            2: f"out[i, {dim} * j + k] = out[i, {dim} * j + k] + vol[0] * f[0, {dim} * j + k]",
+        }[rank]
+        keys = {"f": (function, op2.READ), "vol": (volume, op2.READ), "out": (output, op2.RW)}
+        parloops.par_loop((domain, instructions), dX, keys)
+
+        # Compute patch volume
+        patch_volume = ffunc.Function(functionspace.FunctionSpace(mesh, "CG", 1))
+        domain = "{[i]: 0 <= i < patch.dofs}"
+        instructions = "patch[i] = patch[i] + vol[0]"
+        keys = {"vol": (volume, op2.READ), "patch": (patch_volume, op2.RW)}
+        parloops.par_loop((domain, instructions), dX, keys)
```

Comment:
The patch volume here could become a cached_property so it is not recomputed every time. Maybe we could attach this property to the mesh, so it can be reused more widely across different `Interpolators`.

Reply:
Good idea. I've addressed this in 34ceb670548ad9f7b2385f3337a3ab50991ae3cf, using the `cell_sizes` cached property as a template.